src/main/scala/firrtl/passes/ModuleNamePrefix.scala

```diff
@@ -0,0 +1,57 @@
+package firrtl.passes
+
+import firrtl._
+import firrtl.options.{Dependency, RegisteredTransform, ShellOption}
+import firrtl.transforms.DedupModules
+import firrtl.annotations.NoTargetAnnotation
+import firrtl.stage.RunFirrtlTransformAnnotation
+
+/** Specifies a global prefix for all module names. */
+case class ModuleNamePrefixAnnotation(prefix: String = "") extends NoTargetAnnotation
+
+class ModuleNamePrefixTransform extends Transform with DependencyAPIMigration with RegisteredTransform {
+
+  override def prerequisites = Seq(Dependency[DedupModules])
+  override def optionalPrerequisiteOf = Seq(Dependency[VerilogEmitter])
+
+  override def invalidates(a: Transform) = false
+
+  val options = Seq(
+    new ShellOption[String](
+      longOption = "module-name-prefix",
+      toAnnotationSeq = (a: String) =>
+        Seq(ModuleNamePrefixAnnotation(a), RunFirrtlTransformAnnotation(new ModuleNamePrefixTransform)),
+      helpText = "Add global prefix to every verilog module (default: \"\")",
+      shortOption = Some("prefix"),
+      helpValueName = Some("<prefix>")
+    )
+  )
+
+  private def onStmt(s: ir.Statement, prefix: String): ir.Statement = s match {
+    case i : ir.DefInstance => i.copy(module = prefix + i.module)
+    case other => other.mapStmt(onStmt(_, prefix))
+  }
```

Comment:
Annoyingly the instance cannot be blindly renamed here. Consider the situation of:

```
circuit Foo:
  module Bar:
    skip
  module Prefix_Bar
    skip
  module Foo:
    inst bar1 of Bar
    inst bar2 of Prefix_Bar
```

This will need to track what previous modules have been renamed in some datastructure (a `RenameMap` would work) and then query that to know how to rename the instances.

Reply:
Why is that the case? This transform is written with the assumption that _every_ module gets prefixed. So in this case you would just end up with a circuit like this:

```
circuit Prefix_Foo:
  module Prefix_Bar:
    skip
  module Prefix_Prefix_Bar
    skip
  module Prefix_Foo:
    inst bar1 of Prefix_Bar
    inst bar2 of Prefix_Prefix_Bar
```

Reply:
Ack, you're right.

If the transform is blindly doing the prefixing and cannot deviate from it, then it's not an error.

While extreme scope creep, it would be ideal if a transform like this was idempotent. That would require a formulation like what I'm proposing where the instance cannot be renamed blindly.

Reply:
If we remove the prefix annotation after applying the prefixing and just do nothing if there is no annotation, it would become idempotent, right?

Reply:
Fair. It's idempotent from that perspective. However, the intent of this may be that every Verilog module will have a prefix. If a later transform adds a module, that is an invalidation of module prefixing which can't be enforced by re-running with this approach. It's a definition of what this transform actually means. Is it a static application of a prefix or is it a guarantee that modules have a prefix in Verilog emission?

Reply:
To clarify: I'm good with this as blindly renaming. This would fill an API hole. If we need something more, we can update it in the future.